### PR TITLE
[BUG] Fix forecaster based imputation strategy `Imputer` if forecaster requires `fh` in `fit`

### LINKED ISSUE #4997: [Bug] Imputer with forecaster not working
**Describe the bug**
I know it's more of a question, but I don't understand how I can use Imputer with a forecaster.

```python
from sklearn.ensemble import GradientBoostingRegressor
from sktime.datasets import load_solar
import pandas as pd
import numpy as np
from sktime.transformations.series.impute import Imputer

y = load_solar()
y = y.groupby(by=pd.Grouper(level=0, freq="D")).sum()
y.iloc[100:110] = np.nan

forecaster = make_reduction(GradientBoostingRegressor(), strategy="multioutput")
impute = Imputer(method="forecaster", forecaster=forecaster)
impute.fit_transform(y)
```

In run the code above and I get 

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[14], line 14
     12 forecaster = make_reduction(GradientBoostingRegressor(), strategy="multioutput")
     13 impute = Imputer(method="forecaster", forecaster=forecaster)
---> 14 impute.fit_transform(y)

File [c:\ProgramData\Miniconda3\envs\DLL_Forecast\Lib\site-packages\sktime\transformations\base.py:619](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/Lib/site-packages/sktime/transformations/base.py:619), in BaseTransformer.fit_transform(self, X, y)
    554 """Fit to data, then transform it.
    555 
    556 Fits the transformer to X and y and returns a transformed version of X.
   (...)
    615         Example: i-th instance of the output is the i-th window running over `X`
    616 """
    617 # Non-optimized default implementation; override when a better
    618 # method is possible for a given algorithm.
--> 619 return self.fit(X, y).transform(X, y)

File [c:\ProgramData\Miniconda3\envs\DLL_Forecast\Lib\site-packages\sktime\transformations\base.py:533](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/Lib/site-packages/sktime/transformations/base.py:533), in BaseTransformer.transform(self, X, y)
    530 X_inner, y_inner, metadata = self._check_X_y(X=X, y=y, return_metadata=True)
    532 if not isinstance(X_inner, VectorizedDF):
--> 533     Xt = self._transform(X=X_inner, y=y_inner)
    534 else:
    535     # otherwise we call the vectorized version of predict
    536     Xt = self._vectorize("transform", X=X_inner, y=y_inner)

File [c:\ProgramData\Miniconda3\envs\DLL_Forecast\Lib\site-packages\sktime\transformations\series\impute.py:239](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/Lib/site-packages/sktime/transformations/series/impute.py:239), in Imputer._transform(self, X, y)
    237     X = self._impute_with_forecaster(X, y)
    238 elif self.method == "forecaster":
--> 239     X = self._impute_with_forecaster(X, y)
    240 elif self.method == "mean":
    241     return X.fillna(value=self._mean)

File [c:\ProgramData\Miniconda3\envs\DLL_Forecast\Lib\site-packages\sktime\transformations\series\impute.py:343](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/Lib/site-packages/sktime/transformations/series/impute.py:343), in Imputer._impute_with_forecaster(self, X, y)
    339 fh = ForecastingHorizon(values=na_index, is_relative=False)
    341 # fill NaN before fitting with ffill and backfill (heuristic)
--> 343 self._forecaster.fit(
    344     y=self._X[col].fillna(method="ffill").fillna(method="backfill"),
    345     X=self._y[col].fillna(method="ffill").fillna(method="backfill")
    346     if self._y is not None
    347     else None,
    348 )
    350 # replace missing values with predicted values
    351 X[col][na_index] = self._forecaster.predict(fh=fh, X=y)

File [c:\ProgramData\Miniconda3\envs\DLL_Forecast\Lib\site-packages\sktime\forecasting\base\_base.py:341](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/Lib/site-packages/sktime/forecasting/base/_base.py:341), in BaseForecaster.fit(self, y, X, fh)
    338 self._update_y_X(y_inner, X_inner)
    340 # check forecasting horizon and coerce to ForecastingHorizon object
--> 341 fh = self._check_fh(fh)
    343 # checks and conversions complete, pass to inner fit
    344 #####################################################
    345 vectorization_needed = isinstance(y_inner, VectorizedDF)

File [c:\ProgramData\Miniconda3\envs\DLL_Forecast\Lib\site-packages\sktime\forecasting\base\_base.py:1758](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/Lib/site-packages/sktime/forecasting/base/_base.py:1758), in BaseForecaster._check_fh(self, fh)
   1751         # in case C. fh is not optional in fit: this is fine
   1752         # any error would have already been caught in fit
   1753 
   1754     # A. strategy not fitted (call of fit)
   1755     elif requires_fh:
   1756         # in case fh is not optional in fit:
   1757         # fh must be passed in fit
-> 1758         raise ValueError(
   1759             "The forecasting horizon `fh` must be passed to "
   1760             f"`fit` of {self.__class__.__name__}, but none was found. " + msg
   1761         )
   1762         # in case C. fh is optional in fit:
   1763         # this is fine, nothing to check/raise
   1764 
   1765 # B. fh is passed
   1766 else:
   1767     # If fh is passed, coerce to ForecastingHorizon and validate (all cases)
   1768     fh = check_fh(fh=fh, freq=self._cutoff)

ValueError: The forecasting horizon `fh` must be passed to `fit` of MultioutputTabularRegressionForecaster, but none was found. This is because fitting of the forecaster MultioutputTabularRegressionForecaster depends on `fh`.
```

**Expected behavior**
As far as I understand or what I would expect : 

- The Imputer only forecasts the values missing within the training data, so fh is kind-of a duplicate or unneeded information.
- I can't even pass a fh as it's not allowed in any of the methods.

Am I doing something wrong or missing something?
I couldn't find any example here or online, so that's why I am asking.

**Versions**
<details>

System:
    python: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 17:59:51) [MSC v.1935 64 bit (AMD64)]
executable: [c:\ProgramData\Miniconda3\envs\DLL_Forecast\python.exe](file:///C:/ProgramData/Miniconda3/envs/DLL_Forecast/python.exe)
   machine: Windows-10-10.0.19042-SP0

Python dependencies:
          pip: 23.2.1
       sktime: 0.21.0
      sklearn: 1.3.0
       skbase: 0.4.6
        numpy: 1.24.4
        scipy: 1.11.1
       pandas: 2.0.3
   matplotlib: 3.7.2
       joblib: 1.3.0
  statsmodels: 0.14.0
        numba: 0.57.1
     pmdarima: None
      tsfresh: None
   tensorflow: None
tensorflow_probability: None

</details>

### PR DESCRIPTION
**Pull request by MCRE-BE**
#### Reference Issues/PRs

Fixes #4997

#### What does this implement/fix? Explain your changes.
Passes the missing ForecastingHorizon to the fit method
Adds a test to check with a forecaster requiring fh in fit and not only in predict.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### Did you add any tests for the change?
Yes

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by MCRE-BE** [comment]
@fkiraly : Any advice on the forecaster to use ? Based on docstring alone I can't determine which one to use 😨

**Comment by fkiraly** [comment]
You can find all forecasters that don't require `fh` in `fit` as follows

```python
from sktime.registry import all_estimators

all_estimators(filter_tags={"requires-fh-in-fit": False}, as_dataframe=True)
```

It looks like these are all composites - so the `sklearn` reduction forecaster looks like a good choice? `sklearn` is a core dependency, so we do not introduce additional dependencies in the test, from using your example.

**Comment by fkiraly** [comment]
So, this one should be ok: `forecaster = make_reduction(GradientBoostingRegressor(), strategy="multioutput")`

For better runtime, perhaps replace with linear regressio?n

**Comment by MCRE-BE** [comment]
@fkiraly : Done. Test is added.
In theory, should be ok for testing.

**Comment by fkiraly** [comment]
Linting is failing - the long version on how to set it up locally is here: https://www.sktime.net/en/latest/developer_guide/coding_standards.html

You can also click on "details" (at the right of this:
image
) to see what's wrong

**Review by fkiraly — APPROVED** [review]
That was quick :-)

looks good now!

**Comment by MCRE-BE** [comment]
> Linting is failing - the long version on how to set it up locally is here: https://www.sktime.net/en/latest/developer_guide/coding_standards.html
> 
> You can also click on "details" (at the right of this: image ) to see what's wrong

Thanks. I saw. I added a whitespace after the docstring. Fixed.

I didn't bother installing a new environment just for this small fix. So I don't have all the same linting. Sorry.

**Comment by fkiraly** [comment]
>I didn't bother installing a new environment just for this small fix.

No worries, that's understandable.

File: .all-contributorsrc [hunk 1]
```diff
@@ -2276,6 +2276,15 @@
...
         "doc"
       ]
+    },
+    {
+      "login": "MCRE-BE",
+      "name": "Mathias Creemers",
+      "avatar_url": "https://avatars.githubusercontent.com/u/99316631",
+      "profile": "https://github.com/MCRE-BE",
+      "contributions": [
+        "bug"
```
Comment (fkiraly) [inline .all-contributorsrc:2286]: you can also get the `code` badge, because you also fixed the bug

**Comment by MCRE-BE** [comment]
I'll add it tomorrow. PC is turned off 😜

**Comment by MCRE-BE** [comment]
Done.

**Comment by MCRE-BE** [comment]
It seems all fails are because of urllib...

**Review by fkiraly — APPROVED** [review]

**Pull request merged**